### PR TITLE
Create initial contracts and methods for HTTP device onboarding

### PR DESCRIPTION
**Pull request by timstewartm**
Work in progress

### CODE REVIEW

File: discovery/device/samples/how to guides/TpmSample/Program.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using CommandLine;
+using Microsoft.Azure.Devices.Provisioning.Security;
+using System;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Provisioning.Client.Samples
```
Comment (kagoyal) [inline discovery/device/samples/how to guides/TpmSample/Program.cs]: namespace Microsoft.Azure.Devices.Discovery.Client.Samples

File: discovery/device/samples/how to guides/TpmSample/DiscoveryDeviceClientSample.cs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+using Microsoft.Azure.Devices.Provisioning.Security;
+using Microsoft.Azure.Devices.Shared;
+using System;
+using System.Security.Cryptography.X509Certificates;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.Provisioning.Client.Samples
+{
+    /// <summary>
+    /// Demonstrates how to register a device with the device provisioning service using a certificate, and then
```
Comment (kagoyal) [inline discovery/device/samples/how to guides/TpmSample/DiscoveryDeviceClientSample.cs]: fix comment

File: discovery/device/src/DiscoveryDeviceClient.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Common;
+using Microsoft.Azure.Devices.Discovery.Client.Transport;
+using Microsoft.Azure.Devices.Shared;
+using Newtonsoft.Json;
+
+namespace Microsoft.Azure.Devices.Discovery.Client
+{
+    /// <summary>
+    /// Allows devices to use the Device Provisioning Service.
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryDeviceClient.cs]: Discovery

File: discovery/device/src/DiscoveryDeviceClient.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    /// </summary>
+    public class DiscoveryDeviceClient
+    {
+        private readonly string _globalDeviceEndpoint;
+        private readonly DiscoveryTransportHandler _transport;
+        private readonly SecurityProvider _security;
+
+        /// <summary>
+        /// Creates an instance of the Device Provisioning Client.
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryDeviceClient.cs]: Please fix all occurrences of provisioning service in comments

File: discovery/device/src/DiscoveryDeviceClient.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+        /// <param name="globalDeviceEndpoint">The GlobalDeviceEndpoint for the Device Provisioning Service.</param>
+        /// <param name="securityProvider">The security provider instance.</param>
+        /// <param name="transport">The type of transport (e.g. HTTP, AMQP, MQTT).</param>
+        /// <returns>An instance of the ProvisioningDeviceClient</returns>
+        public static DiscoveryDeviceClient Create(
+            string globalDeviceEndpoint,
+            SecurityProvider securityProvider,
+            DiscoveryTransportHandler transport)
+        {
+            return new DiscoveryDeviceClient(globalDeviceEndpoint, securityProvider, transport);
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryDeviceClient.cs:34]: As only TPM security provider is allowed right now, let's put the check here when creating client

File: discovery/device/src/DiscoveryDeviceClient.cs [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+            Logging.Associate(this, _transport);
+        }
+
+        /// <summary>
+        /// Stores product information that will be appended to the user agent string that is sent to IoT hub.
+        /// </summary>
+        public string ProductInfo { get; set; }
+
+        /// <summary>
+        /// Gets challenge string to decrypt to onboard device
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryDeviceClient.cs]: nit: Issues challenge to authenticate the given device

File: discovery/device/src/DiscoveryDeviceClient.cs [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryDeviceClient.cs]: returns nonce challenge

File: discovery/device/src/DiscoveryTransportException.cs [hunk 8]
```diff
@@ -0,0 +1,133 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.Serialization;
+
+namespace Microsoft.Azure.Devices.Discovery.Client
+{
+    /// <summary>
+    /// Represents errors reported by the Provisioning Transport Handlers.
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryTransportException.cs]: Please fix all comments - Discovery Transport Handlers

File: discovery/device/src/DiscoveryTransportException.cs [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+    /// </summary>
+    public class DiscoveryTransportException : Exception
+    {
+        private const string IsTransientValueSerializationStoreName = "ProvisioningTransportException-IsTransient";
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryTransportException.cs]: DiscoveryTransportException

File: common/src/Logging.ProvisioningTransport.Common.cs [hunk 10]
```diff
@@ -57,6 +58,38 @@ internal sealed partial class Logging : EventSource
                     retryAfterSeconds,
                     status);
 
+        [NonEvent]
+        public static void OnboardDevice(
```
Comment (kagoyal) [inline common/src/Logging.ProvisioningTransport.Common.cs:62]: Why do we need this change? This is also inside provisioning amqp layer? Is file just wrongly named?

File: discovery/device/src/DiscoveryDeviceClient.cs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+                return _transport.IssueChallengeAsync(request, cancellationToken);
+            }
+            else
+            {
+                throw new NotSupportedException($"{nameof(_security)} must be of type {nameof(SecurityProviderTpm)}");
+            }
+        }
+
+        /// <summary>
+        /// Gets challenge string to decrypt to onboard device
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryDeviceClient.cs]: fix comment

File: discovery/device/src/DiscoveryDeviceClient.cs [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        /// </summary>
+        /// <param name="nonce"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="NotSupportedException">Security provider must be TPM</exception>
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryDeviceClient.cs]: Let's fail when creating client itself as mentioned above

File: discovery/device/src/DiscoveryTransportHandler.cs [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+        /// <summary>
+        /// Gets or sets the proxy for Discovery Client operations.
+        /// </summary>
+        public IWebProxy Proxy { get; set; }
+
+        /// <summary>
+        /// A callback for remote certificate validation.
+        /// If incorrectly implemented, your device may fail to connect to DPS and/or be open to security vulnerabilities.
+        /// </summary>
+        public RemoteCertificateValidationCallback RemoteCertificateValidationCallback { get; set; }
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryTransportHandler.cs]: not used, remove this

File: discovery/device/src/OnboardingInfo.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Security.Cryptography.X509Certificates;
+
+namespace Microsoft.Azure.Devices.Discovery.Client
+{
+    /// <summary>
+    /// Holds information returned from the getOnboardingInfo endpoint
```
Comment (kagoyal) [inline discovery/device/src/OnboardingInfo.cs]: /// Holds information returned from the getOnboardingInfo **request**

File: discovery/device/src/OnboardingInfo.cs [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+    /// </summary>
+    public class OnboardingInfo
+    {
+        /// <summary>
+        /// The endpoint to reach the provisioning service
+        /// </summary>
+        public string EdgeProvisioningEndpoint { get; private set; }
+        /// <summary>
```
Comment (kagoyal) [inline discovery/device/src/OnboardingInfo.cs:18]: add new line before summary

File: discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Devices.Common;
+using Microsoft.Azure.Devices.Discovery.Client.Transport.Http;
+using Microsoft.Azure.Devices.Discovery.Client.Transport.Http.Models;
+using Microsoft.Azure.Devices.Provisioning.Client.Transport;
+using Microsoft.Azure.Devices.Shared;
+using Microsoft.Rest;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
```
Comment (kagoyal) [inline discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs]: remove all unused usings

File: discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+            try
+            {
+                var securityProvider = request.Security;
+
+                using var httpClientHandler = new HttpClientHandler()
+                {
+                    // Cannot specify a specific protocol here, as desired due to an error:
+                    //   ProvisioningDeviceClient_ValidRegistrationId_AmqpWithProxy_SymmetricKey_RegisterOk_GroupEnrollment failing for me with System.PlatformNotSupportedException: Operation is not supported on this platform.	
+                    // When revisiting TLS12 work for DPS, we should figure out why. Perhaps the service needs to support it.	
+
```
Comment (kagoyal) [inline discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs]: remove the comment here

File: discovery/device/src/DiscoveryTransportHandler.cs [hunk 18]
```diff
@@ -0,0 +1,114 @@
...
+
+                _port = value;
+            }
+        }
+        
+        /// <summary>
+        /// Issue challenge
+        /// </summary>
+        /// <returns></returns>
+        public virtual Task<string> IssueChallengeAsync(DiscoveryTransportIssueChallengeRequest request, CancellationToken cancellationToken)
```
Comment (kagoyal) [inline discovery/device/src/DiscoveryTransportHandler.cs:74]: Please copy summary from generated code - 
            /// Issue TPM Challenge. Will return an encrypted nonce, that can be used to
            /// sign a SAS Token for the GetOnboardingInfo request.

File: discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs [hunk 19]
```diff
@@ -0,0 +1,276 @@
...
+                throw new DiscoveryTransportException($"HTTP transport exception", ex, true);
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, $"{nameof(DiscoveryTransportHandlerHttp)}.{nameof(GetOnboardingInfoAsync)}");
+            }
+        }
+
+        private string GenerateCSR(RSA rsa, string deviceId)
```
Comment (kagoyal) [inline discovery/transport/http/src/DiscoveryTransportHandlerHttp.cs:255]: please add summary

File: discovery/transport/http/src/Logging.DiscoveryTransport.Http.cs [hunk 20]
```diff
@@ -0,0 +1,63 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics.Tracing;
+
+namespace Microsoft.Azure.Devices.Shared
+{
+    // d209b8a1-2e02-5724-f341-677227d0ed22
```
Comment (kagoyal) [inline discovery/transport/http/src/Logging.DiscoveryTransport.Http.cs]: can you check if these GUIDs have to be new/different for discovery.
Also let's test we are getting logs once logging is enabled

File: e2e/test/E2ETests.csproj [hunk 21]
```diff
@@ -1,7 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks Condition="'$(OS)' == 'Windows_NT'">net6.0;net5.0;netcoreapp3.1;netcoreapp2.1.30;net472;net451</TargetFrameworks>
```
Comment (kagoyal) [inline e2e/test/E2ETests.csproj:3]: Let's not change this condition in e2etests project

File: e2e/test/config/TestConfiguration.Discovery.cs [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+using System.Security.Cryptography.X509Certificates;
+using FluentAssertions;
+
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    public static partial class TestConfiguration
+    {
+        public static partial class Discovery
+        {
+            public static string ConnectionString => GetValue("PROVISIONING_CONNECTION_STRING");
```
Comment (kagoyal) [inline e2e/test/config/TestConfiguration.Discovery.cs]: DISCOVERY_CONNECTION_STRING?

File: provisioning/device/src/CommonExtensions.cs [hunk 23]
```diff
@@ -4,9 +4,12 @@
 using System;
 using System.Text.RegularExpressions;
 
-namespace Microsoft.Azure.Devices.Provisioning.Client.Extensions
+namespace Microsoft.Azure.Devices.Provisioning.Client
```
Comment (kagoyal) [inline provisioning/device/src/CommonExtensions.cs]: why this namespace change?
Let's not change it

File: provisioning/device/src/DeviceOnboardingResult.cs [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+        /// <param name="onboardingStatus"> The status of the onboarding process. </param>
+        public Device(string registrationId, string onboardingStatus)
+        {
+            RegistrationId = registrationId;
+            OnboardingStatus = onboardingStatus;
+        }
+
+        /// <summary> The registrationId. </summary>
+        public string RegistrationId { get; }
+        /// <summary> The status of the onboarding process.Possible values include:
```
Comment (kagoyal) [inline provisioning/device/src/DeviceOnboardingResult.cs:50]: newline after this

**Review by kagoyal — CHANGES_REQUESTED** [review]

File: e2e/test/E2ETests.csproj (hunk 21 above)
Comment (timstewartm) [reply e2e/test/E2ETests.csproj:3]: The discovery code is not compatible with the other frameworks.

Comment (timstewartm) [reply e2e/test/E2ETests.csproj:3]: Will evaluate dependencies and take this in future PR

**Review by kagoyal — APPROVED** [review]

**Pull request merged**